schema: Apply `sstable_compression_user_table_options` to CQL aux and Alternator tables

ndragazis opened the pull request:
In PR 5b6570be52 we introduced the config option `sstable_compression_user_table_options` to allow adjusting the default compression settings for user tables. However, the new option was hooked into the CQL layer and applied only to CQL base tables, not to the whole spectrum of user tables: CQL auxiliary tables (materialized views, secondary indexes, CDC log tables), Alternator base tables, Alternator auxiliary tables (GSIs, LSIs, Streams).

This gap also led to inconsistent default compression algorithms after we changed the option’s default algorithm from LZ4 to LZ4WithDicts (adf9c426c2).

This series introduces a general “schema initializer” mechanism in `schema_builder` and uses it to apply the default compression settings uniformly across all user tables. This ensures that all base and aux tables take their default compression settings from config.

Fixes #26914.

Backport justification: LZ4WithDicts is the new default since 2025.4, but the config option exists since 2025.2. Based on severity, I suggest we backport only to 2025.4 to maintain consistency of the defaults.

nyh commented on test/cluster/test_sstable_compression_config.py:
```diff
@@ -145,4 +147,159 @@ async def create_table_and_check_compression(cql, keyspace, table_name, expected
+    algorithm hardcoded to LZ4Compressor.
+    """
+    # Start Scylla with custom compression options
+    alternator_address = '0.0.0.0'
```
What's this 0.0.0.0?

nyh commented on schema/schema_builder.hh:
```diff
@@ -52,6 +54,7 @@ public:
     schema_builder(const schema_ptr);
 
     static int register_static_configurator(static_configurator&& configurator);
+    static int register_schema_initializer(schema_initializer&& initializer);
```
I think you should add a comment here or in the header file explaining what a "static configurator does", what a "schema initializer" does - and when to use each. To be honest, I didn't even realize until a minute ago we even have this concept of "static configurator" here, and I still wonder if we need both.

ndragazis replied:
Comments added in v2.

ndragazis commented on test/cluster/test_sstable_compression_config.py:
```diff
@@ -145,4 +147,159 @@ async def create_table_and_check_compression(cql, keyspace, table_name, expected
+    await create_table_and_check_compression(cql, "test_ks", "table_after_upgrade", "LZ4WithDictsCompressor", "after upgrade and feature enabled")
+
+
+# Copied from test.cluster.test_alternator
```
Yes, I just didn't want to introduce a dependency between test files. Ideally, it should become a utility function.

ndragazis commented on test/cluster/test_sstable_compression_config.py:
```diff
@@ -6,9 +6,11 @@
 
 import os
 import time
+import boto3
```
I think it's the other way around: the new tests are needed (they test a stronger condition than the old tests), and they make the old tests redundant. Let me explain:

The existing tests check that the default compression settings are the same for all user tables. They don't care about where the defaults come from. As of this patch, they should come from the default value of `sstable_compression_user_table_options`. But this is an implementation detail for these tests.

The new tests check that, if you explicitly set a value for `sstable_compression_user_table_options`, this will be applied as the default to all tables. So, they verify the actual source of the defaults, not the value per se.

ndragazis commented:
v3 Changelog:
* Added the schema initializer to scylla-sstable as well (to fix the failing test).

The dtests are failing because some MV tests still expect the old default. I'll send a separate PR for that.

ndragazis commented on schema/schema.cc:
```diff
@@ -1692,6 +1698,28 @@ int schema_builder::register_static_configurator(static_configurator&& configura
     return 0;
 }
 
+auto schema_builder::schema_initializers() -> std::vector<schema_initializer>& {
```
I wanted to be consistent with the syntax used for the pre-existing `schema_builder::static_configurators()`.

I'll change it.

ndragazis commented on main.cc:
```diff
@@ -824,6 +826,16 @@ To start the scylla server proper, simply invoke as: scylla server (or just scyl
+            // * Needs to be done after the initialization of the configurables (this particular initializer depends on `db::extensions::is_extension_internal_keyspace()`).
+            // * Needs to be notified about when `sstable_compression_user_table_options` is properly initialized so that it can start using it.
+            //   This is done via an atomic flag which is set to true once the configuration is ready.
+            std::atomic<bool> compression_config_ready{false};
```
> I think you should follow the way other configuration items which depend on features do this.

OK, so you are saying I should not be modifying the config option's default value depending on the feature flag, but rather define an accessor that takes the feature flag into account and returns the correct value in any case. And this is what we do for the sstable format (the accessor is the `get_preferred_sstable_version()`).

Yeah, I guess we can do that.

I'll get back to it later this week, once I finish my current task.

ndragazis commented:
v6 Changelog:
* Rebased
* Added a new patch: "db: config: Add accessor for sstable_compression_user_table_options"
   The patch replaces the in-place updates of the config option (for falling back to non-dict until the dictionary compression feature is enabled), with an accessor that returns the correct options, depending on the feature flag.
   As a result, the atomic variable was also removed.
   Addresses: https://github.com/scylladb/scylladb/pull/27204#discussion_r2639672889

<details>
<summary>Full v5 -> v6 intediff:</summary>

```diff
diff --git a/db/config.cc b/db/config.cc
index 31c0fccb38..fb0fe06368 100644
--- a/db/config.cc
+++ b/db/config.cc
@@ -1785,6 +1785,21 @@ const db::extensions& db::config::extensions() const {
     return *_extensions;
 }
 
+compression_parameters db::config::get_sstable_compression_user_table_options(bool dicts_feature_enabled) const {
+    if (sstable_compression_user_table_options.is_set()
+            || dicts_feature_enabled
+            || !sstable_compression_user_table_options().uses_dictionary_compressor()) {
+        return sstable_compression_user_table_options();
+    } else {
+        // Fall back to non-dict if dictionary compression is not enabled cluster-wide.
+        auto options = sstable_compression_user_table_options();
+        auto params = options.get_options();
+        auto algo = compression_parameters::non_dict_equivalent(options.get_algorithm());
+        params[compression_parameters::SSTABLE_COMPRESSION] = sstring(compression_parameters::algorithm_to_name(algo));
+        return compression_parameters{params};
+    }
+}
+
 std::map<sstring, db::experimental_features_t::feature> db::experimental_features_t::map() {
     // We decided against using the construct-on-first-use idiom here:
     // https://github.com/scylladb/scylla/pull/5369#discussion_r353614807
diff --git a/db/config.hh b/db/config.hh
index fd50bf8354..8d8d0d888c 100644
--- a/db/config.hh
+++ b/db/config.hh
@@ -419,7 +419,13 @@ class config final : public utils::config_file {
     named_value<bool> enable_sstables_mc_format;
     named_value<bool> enable_sstables_md_format;
     named_value<sstring> sstable_format;
+
+    // NOTE: Do not use this option directly.
+    // Use get_sstable_compression_user_table_options() instead.
     named_value<compression_parameters> sstable_compression_user_table_options;
+
+    compression_parameters get_sstable_compression_user_table_options(bool dicts_feature_enabled) const;
+
     named_value<bool> sstable_compression_dictionaries_allow_in_ddl;
     named_value<bool> sstable_compression_dictionaries_enable_writing;
     named_value<float> sstable_compression_dictionaries_memory_budget_fraction;
diff --git a/main.cc b/main.cc
index 7f6fa19050..c3a4308a3f 100644
--- a/main.cc
+++ b/main.cc
@@ -802,7 +802,6 @@ To start the scylla server proper, simply invoke as: scylla server (or just scyl
             }).get();
             ::stop_signal stop_signal; // we can move this earlier to support SIGINT during initialization
             read_config(opts, *cfg).get();
-
 #ifdef SCYLLA_ENABLE_ERROR_INJECTION
             enable_initial_error_injections(*cfg).get();
 #endif
@@ -816,12 +815,14 @@ To start the scylla server proper, simply invoke as: scylla server (or just scyl
 
             // Register schema initializer for default compression settings.
             // This statement is deliberately positioned here:
-            // * Needs to be done before starting services that use `schema_builder`s to make sure that all `schema_builder`s will use it.
-            // * Needs to be done after the initialization of the configurables (this particular initializer depends on `db::extensions::is_extension_internal_keyspace()`).
-            // * Needs to be notified about when `sstable_compression_user_table_options` is properly initialized so that it can start using it.
-            //   This is done via an atomic flag which is set to true once the configuration is ready.
-            std::atomic<bool> compression_config_ready{false};
-            register_compression_initializer(*cfg, &compression_config_ready);
+            // * Needs to be done before starting services that use `schema_builder`s
+            //   because registration is not thread-safe (initializer vector is
+            //   not thread-local) and every `schema_builder` must use it.
+            // * Needs to be done after the initialization of the configurables
+            //   because this particular initializer depends on `db::extensions::is_extension_internal_keyspace()`.
+            register_compression_initializer(*cfg, [&feature_service] {
+                return bool(feature_service.local().sstable_compression_dicts);
+            });
 
             cfg->setup_directories();
 
@@ -2231,54 +2232,16 @@ To start the scylla server proper, simply invoke as: scylla server (or just scyl
             // Semantic validation of sstable compression parameters from config.
             // Adding here (i.e., after `join_cluster`) to ensure that the
             // required SSTABLE_COMPRESSION_DICTS cluster feature has been negotiated.
-            //
-            // Also, if the dictionary compression feature is not enabled, use
-            // LZ4Compressor as the default algorithm instead of LZ4WithDictsCompressor.
-            const auto& dicts_feature_enabled = feature_service.local().sstable_compression_dicts;
-            auto& sstable_compression_options = cfg->sstable_compression_user_table_options;
-
-            gms::feature::listener_registration reg_listener;
-
-            if (!sstable_compression_options.is_set() && !dicts_feature_enabled) {
-                if (sstable_compression_options().get_algorithm() != compression_parameters::algorithm::lz4_with_dicts) {
-                    on_internal_error(startlog, "expected LZ4WithDictsCompressor as default algorithm for sstable_compression_user_table_options.");
-                }
-
-                startlog.info("SSTABLE_COMPRESSION_DICTS feature is disabled. Overriding default SSTable compression to use LZ4Compressor instead of LZ4WithDictsCompressor.");
-                compression_parameters original_params{sstable_compression_options().get_options()};
-                auto params = sstable_compression_options().get_options();
-                params[compression_parameters::SSTABLE_COMPRESSION] = sstring(compression_parameters::algorithm_to_name(compression_parameters::algorithm::lz4));
-                smp::invoke_on_all([&sstable_compression_options, params = std::move(params)] {
-                    if (!sstable_compression_options.is_set()) { // guard check; in case we ever make the option live updateable
-                        sstable_compression_options(compression_parameters{params}, utils::config_file::config_source::None);
-                    }
-                }).get();
-
-                // Register a callback to update the default compression algorithm when the feature is enabled.
-                // Precondition:
-                //   The callback must run inside seastar::async context:
-                //   - If the listener fires immediately, we are running inside seastar::async already.
-                //   - If the listener is deferred, `feature_service::enable()` runs it inside seastar::async.
-                reg_listener = feature_service.local().sstable_compression_dicts.when_enabled([&sstable_compression_options, params = std::move(original_params)] {
-                    startlog.info("SSTABLE_COMPRESSION_DICTS feature is now enabled. Overriding default SSTable compression to use LZ4WithDictsCompressor.");
-                    smp::invoke_on_all([&sstable_compression_options, params = std::move(params)] {
-                        if (!sstable_compression_options.is_set()) { // guard check; in case we ever make the option live updateable
-                            sstable_compression_options(params, utils::config_file::config_source::None);
-                        }
-                    }).get();
-                });
-            }
+            const auto dicts_feature_enabled = bool(feature_service.local().sstable_compression_dicts);
 
             try {
-                cfg->sstable_compression_user_table_options().validate(
-                        compression_parameters::dicts_feature_enabled(bool(dicts_feature_enabled)));
+                cfg->get_sstable_compression_user_table_options(dicts_feature_enabled).validate(
+                        compression_parameters::dicts_feature_enabled(dicts_feature_enabled));
             } catch (const std::exception& e) {
                 startlog.error("Invalid sstable_compression_user_table_options: {}", e.what());
                 throw bad_configuration_error();
             }
 
-            compression_config_ready.store(true, std::memory_order_release);
-
             dictionary_service dict_service(
                 dict_sampler,
                 sys_ks.local(),
diff --git a/schema/compression_initializer.hh b/schema/compression_initializer.hh
index b127188673..f42788c95f 100644
--- a/schema/compression_initializer.hh
+++ b/schema/compression_initializer.hh
@@ -1,5 +1,5 @@
 /*
- * Copyright (C) 2025-present ScyllaDB
+ * Copyright (C) 2026-present ScyllaDB
  */
 
 /*
@@ -8,7 +8,7 @@
 
 #pragma once
 
-#include <atomic>
+#include <functional>
 
 #include "db/config.hh"
 #include "db/extensions.hh"
@@ -30,18 +30,19 @@ namespace {
  * CQL base tables, materialized views, secondary indexes, CDC log tables,
  * Alternator base tables, Alternator GSIs, Alternator LSIs and Alternator Streams.
  */
-inline void register_compression_initializer(db::config& cfg, std::atomic<bool>* cfg_ready_ptr = nullptr) {
-    schema_builder::register_schema_initializer([&cfg, cfg_ready_ptr](schema_builder& builder) {
-        bool _cfg_ready = cfg_ready_ptr ? cfg_ready_ptr->load(std::memory_order_acquire) : true;
+inline void register_compression_initializer(db::config& cfg, std::function<bool()> dicts_feature_enabled_fn) {
+    schema_builder::register_schema_initializer([&cfg, dicts_feature_enabled_fn = std::move(dicts_feature_enabled_fn)](schema_builder& builder) {
 
-        if (is_internal_keyspace(builder.ks_name())
-                || cfg.extensions().is_extension_internal_keyspace(builder.ks_name())
-                || !_cfg_ready) {
+        if (is_internal_keyspace(builder.ks_name()) || cfg.extensions().is_extension_internal_keyspace(builder.ks_name())) {
             builder.set_compressor_params(compression_parameters::algorithm::lz4);
         } else {
-            builder.set_compressor_params(cfg.sstable_compression_user_table_options());
+            builder.set_compressor_params(cfg.get_sstable_compression_user_table_options(dicts_feature_enabled_fn()));
         }
     });
 }
 
+inline void register_compression_initializer(db::config& cfg, bool dicts_feature_enabled) {
+    register_compression_initializer(cfg, [dicts_feature_enabled] { return dicts_feature_enabled; });
+}
+
 }
\ No newline at end of file
diff --git a/schema/schema.hh b/schema/schema.hh
index c4ae191a7f..4c54f2bc4c 100644
--- a/schema/schema.hh
+++ b/schema/schema.hh
@@ -493,13 +493,13 @@ class schema_extension {
 /**
  * Schema static properties
  *
- * Represents some internal properties not visible to users.
- * Unlike normal schema properties, these properties are NOT persisted in system
- * tables along with the schema (and as such, one could argue they are not even
- * part of the schema, despite their name); they are immutable table properties
- * derived solely from keyspace/table names (thereby "static") and carried as
- * part of in-memory `schema` objects.
- * (https://github.com/scylladb/scylladb/pull/13170#issuecomment-1469848086)
+ * Represents internal table properties not visible to users.
+ * Their values are derived solely from keyspace/table names (thereby "static").
+ * Despite their name, they are NOT part of the schema (i.e., they do not map to
+ * user-visible schema fields) and they are NOT persisted in system tables.
+ * The reason they are carried along with in-memory `schema` objects is just
+ * convenience, see:
+ * https://github.com/scylladb/scylladb/pull/13170#issuecomment-1469848086
  */
 struct schema_static_props {
     bool use_null_sharder = false; // use a sharder that puts everything on shard 0
diff --git a/sstables/compressor.hh b/sstables/compressor.hh
index 6a8607ebcc..794562381f 100644
--- a/sstables/compressor.hh
+++ b/sstables/compressor.hh
@@ -113,12 +113,26 @@ class compression_parameters {
     bool compression_enabled() const { 
         return _algorithm != algorithm::none;
     }
+    bool uses_dictionary_compressor() const {
+        return _algorithm == algorithm::lz4_with_dicts
+            || _algorithm == algorithm::zstd_with_dicts;
+    }
     static compression_parameters no_compression() {
         return compression_parameters(algorithm::none);
     }
     bool operator==(const compression_parameters&) const = default;
     static std::string_view algorithm_to_name(algorithm);
     static std::string algorithm_to_qualified_name(algorithm);
+    static algorithm non_dict_equivalent(algorithm algo) {
+        switch (algo) {
+        case algorithm::lz4_with_dicts:
+            return algorithm::lz4;
+        case algorithm::zstd_with_dicts:
+            return algorithm::zstd;
+        default:
+            return algo;
+        }
+    }
 private:
     static void validate_options(const std::map<sstring, sstring>&);
     static algorithm name_to_algorithm(std::string_view name);
diff --git a/test/boost/cql_query_test.cc b/test/boost/cql_query_test.cc
index 0931749a09..26e0790bd3 100644
--- a/test/boost/cql_query_test.cc
+++ b/test/boost/cql_query_test.cc
@@ -2015,7 +2015,8 @@ SEASTAR_TEST_CASE(test_table_compression) {
 
         e.execute_cql("create table tb6 (foo text PRIMARY KEY, bar text);").get();
         BOOST_REQUIRE(e.local_db().has_schema("ks", "tb6"));
-        BOOST_REQUIRE(e.local_db().find_schema("ks", "tb6")->get_compressor_params().get_algorithm() == e.local_db().get_config().sstable_compression_user_table_options().get_algorithm());
+        const auto dicts_feature_enabled = bool(e.local_db().features().sstable_compression_dicts);
+        BOOST_REQUIRE(e.local_db().find_schema("ks", "tb6")->get_compressor_params().get_algorithm() == e.local_db().get_config().get_sstable_compression_user_table_options(dicts_feature_enabled).get_algorithm());
     });
 }
 
diff --git a/test/lib/cql_test_env.cc b/test/lib/cql_test_env.cc
index 4d58cf1703..fa473eb5f0 100644
--- a/test/lib/cql_test_env.cc
+++ b/test/lib/cql_test_env.cc
@@ -578,7 +578,10 @@ class single_node_cql_env : public cql_test_env {
             auto stop_configurables = defer_verbose_shutdown("configurables", [&] { notify_set.notify_all(configurable::system_state::stopped).get(); });
 
             auto schema_initializer_checkpoint = schema_builder::capture_schema_initializers_checkpoint();
-            register_compression_initializer(*cfg);
+            register_compression_initializer(*cfg, [this] {
+                return bool(_feature_service.local().sstable_compression_dicts);
+            });
+
             // Important to restore schema initializers during shutdown to
             // support tests that repeatedly create `cql_test_env` instances.
             auto unregister_schema_initializers = defer_verbose_shutdown("schema initializers", [schema_initializer_checkpoint] {
diff --git a/tools/scylla-sstable.cc b/tools/scylla-sstable.cc
index cc3dede5de..37545dd27f 100644
--- a/tools/scylla-sstable.cc
+++ b/tools/scylla-sstable.cc
@@ -2790,17 +2790,17 @@ Validate the specified sstables:
 
         // Register schema initializer for default compression settings.
         //
-        // Note that the default compression settings are determined from the
-        // `sstable_compression_user_table_options` in scylla.yaml.
-        // If we fail to locate scylla.yaml, they cannot be determined, so we
-        // assume them to be the default values of the config option.
+        // Note that the initializer determines the default compression settings
+        // from the `sstable_compression_user_table_options` in scylla.yaml.
+        // If we fail to locate scylla.yaml, the initializer uses the default
+        // values of the config option.
         //
-        // Note also that in `scylla_main()` the default compressor in
-        // `sstable_compression_user_table_options` is reverted from LZ4WithDicts
-        // to LZ4 while the SSTABLE_COMPRESSION_DICTS feature flag remains disabled.
-        // However, scylla-sstable has no context about features, so we optimistically
-        // assume that the feature is enabled and make no such adjustments here.
-        register_compression_initializer(dbcfg);
+        // The default values of the config option, and specifically the default
+        // compression algorithm, is not hardcoded; it is either LZ4 or LZ4WithDicts
+        // depending on whether the dictionary compression feature is enabled.
+        // Since scylla-sstable has no context about features, we optimistically
+        // assume that the feature is enabled.
+        register_compression_initializer(dbcfg, true);
 
         {
             unsigned schema_sources = 0;

ndragazis commented:
> > #### Failed Tests (1/129675):
> > 
> > * [snapshot_list_contains_dropped_tables.debug.4](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/23714/testReport/junit/boost.database_test/cc/debug___Unit_Tests_Custom___debug___snapshot_list_contains_dropped_tables_debug_4) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+snapshot_list_contains_dropped_tables.debug.4)
> 
> Flaky test, unrelated to this PR. Opened an issue to track it (#28195).

The issue has been resolved.

@scylladbbot trigger-ci

nyh commented:
CI failure is a recently flaky test https://github.com/scylladb/scylladb/issues/28205.
@scylladbbot trigger-ci

nyh commented on schema/schema.cc:
```diff
@@ -1692,6 +1698,28 @@ int schema_builder::register_static_configurator(static_configurator&& configura
 }
 
+auto schema_builder::schema_initializers() -> std::vector<schema_initializer>& {
+    static std::vector<schema_initializer> result{};
```
‎The threads start running right from the start. But I guess maybe you're doing this from the main() thread, not on every thread. I hope you are following all the details because I couldn't. But it sounds like it can work.

ndragazis replied:
> The threads start running right from the start. But I guess maybe you're doing this from the main() thread, not on every thread. I hope you are following all the details because I couldn't. But it sounds like it can work.

Yes, I'm doing the registration from the main thread. I'm also doing it **before** starting any services, to make sure that this static vector will not be accessed by other threads while it is being modified by the main thread.
